lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,444 @@
+        except IndexError:
+            raise ValueError(
+                f"Could not find key '{bits}' in file permissions dictionary"
+            )
+
+    return fperm
+
+
+class Hierarchy(object):
+    """Super class of Hierarchies"""
+
+    UNEXPECTED_DIRS = "unexpected_dirs"
+    SUBDIR_STATUS_INDICATORS = "subdir_status_indicators"
+    UNEXPECTED_SYMLINKS = "unexpected_symlinks"
+    UNEXPECTED_OBJECTS = "unexpected_objects"
+    NON_PREFIXES = "non_prefixes"
+    WRONG_PREFIXES = "wrong_prefixes"
+    PREFIX_STATUS_INDICATORS = "prefix_status_indicators"
+    MIALIST = "mialist"
+    EMPTY_CONTROLLERS = "empty_controllers"
+    UNEXPECTED_CONTROLLERS = "unexpected_controllers"
+    INVALID_TB_DIRS = "invalid_tb_dirs"
+    EMPTY_TARBALL_DIRS = "empty_tarball_dirs"
+    INVALID_UNPACKING_DIRS = "invalid_unpacking_dirs"
+    TARBALL_LINKS = "tarball_links"
+    INVALID_TB_LINKS = "invalid_tb_links"
+    INCORRECT_TB_DIR_LINKS = "incorrect_tb_dir_links"
+    INVALID_TB_DIR_LINKS = "invalid_tb_dir_links"
+    UNUSED_PREFIX_FILES = "unused_prefix_files"
+    MISSING_PREFIX_FILES = "missing_prefix_files"
+    BAD_PREFIX_FILES = "bad_prefix_files"
+    BAD_PREFIXES = "bad_prefixes"
+    UNEXPECTED_USER_LINKS = "unexpected_user_links"
+    WRONG_USER_LINKS = "wrong_user_links"
+    INCONCEIVABLE = "inconceivable"
+    ERROR = "error"
+
+    def __init__(self, name, path, config):
+        self.name = name
+        self.path = path
+        self.config = config
+        self.controllers = list()
+        self.bad_controllers = list()
+        self.validation_list = None
+
+    def add_controller(self, controller):
+        self.controllers.append(controller)
+
+    def add_bad_controller(self, controller):
+        self.bad_controllers.append(controller)
+
+    def add_error_or_inconceivable_entries(self, val, controller, dirs):
+        self.validation_list[val][controller] = dirs
+
+    def add_unexpected_entries(self, val, controller, dirs):
+        """Gather values for each validation_list dict key"""
+        self.validation_list[val][controller] = dirs
+
+    def header(self, fp, status):
+        """Adds main starting and ending messages for each Hierarchy"""
+        lead_newline = "\n" if status == "start" else ""
+        fp.write(
+            f"{lead_newline}\n{status}-{self.config.TS[4:]}: {self.name} hierarchy: {self.path}\n"
+        )
+        return
+
+    def check_controller_in_list(self, controller):
+        """Checks whether there are controllers in dictionary or not"""
+        return any(
+            [
+                controller in self.validation_list[key]._dict
+                for key in self.validation_list
+            ]
+        )
+
+    """
+    ##FIXME: The mention of hierarchies(archive, controller) from here on, in
+            Hierarchy class is mainly to keep the format algned with the shell script
+    """
+
+    def dump_check(self, fp, controller, key, cnt, hierarchy=False):
```

Comment:
Doesn't 'hierarchy=False' mean at the time of calling this function signature, providing the 'hierarchy' attribute is not important? if you do not provide a hierarchy it will not throw an error.

Reply:
I wanted to provide an optional value. I will have to look into that

Reply:
After refactoring and removing the lead_tab module. it straightened out as str.